File: Arrays_and_Hashing/49_GroupAnagrams/groupAnagram.dart
List<List<String>> groupAnagram(List strs) {
  Map<String, List<String>> res = {};
  for (String s in strs) {
    List<int> count = List.filled(26, 0);
    for (int i = 0; i < s.length; i++) {
      count[s.codeUnitAt(i) - 'a'.codeUnitAt(0)] += 1;
    }
    String key = count.toString();
    if (!res.containsKey(key)) {
      res[key] = [];
    }
    res[key]!.add(s);
  }
  return res.values.toList();
}

void main(List<String> args) {
  print(groupAnagram(["eat", "tea", "tan", "ate", "nat", "bat"]));
}

File: Arrays_and_Hashing/49_GroupAnagrams/group_anagram.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(groupAnagram([]string{"eat", "tea", "tan", "ate", "nat", "bat"}))
}

func groupAnagram(strs []string) [][]string {
	var (
		key      string
		anagram  map[string][]string = make(map[string][]string, 0)
		alphabet [26]int
		ans      [][]string = make([][]string, 0, len(anagram))
	)
	for _, v := range strs {
		for _, ele := range v {
			alphabet[ele-'a']++
		}
		key = fmt.Sprint(alphabet)
		anagram[key] = append(anagram[key], v)
		alphabet = [26]int{}
	}
	for _, val := range anagram {
		ans = append(ans, val)
	}
	return ans
}
